sherpa/models/model.py

```diff
@@ -121,16 +169,64 @@ def __str__(self):
     def __iter__(self):
         return iter([self])
 
-    # Make parameter access case insensitive
     def __getattr__(self, name):
```

Comment:
There is a lot of complexity in this method. At the very least, I would suggest it be broken down into smaller pieces to aid readability and maintainability. On the other hand, wouldn't it be much easier to just build a dictionary of key value pairs upon instantiation? The keys would be the acceptable (lower case) names and the values would be the parameters, then `__getattr__` would simply be a lookup in the dictionary. If it is safer to defer this action after the instantiation, for whatever reason, I would create the map the first time `__getattr__` is called and use it during subsequent calls.

Even better, why not overriding `__setattr__` to add `Parameter`s to an index when they are added in the initializer? I would find this closer to the design I would have expected for parameters all along, i.e. to be declared as class members (with instance values) rather than instance members. In any case, I would then implement a method for declaring alternative names, which takes the alternative name and the reference to the parameter itself as arguments. I am less concerned about the mechanism by which alternative names are registered to trigger a warning, which is more of an implementation detail.